pkg/ipamd/rpc_handler.go

```diff
@@ -204,11 +204,11 @@ func (s *server) DelNetwork(ctx context.Context, in *rpc.DelNetworkRequest) (*rp
 	eni, ip, deviceNumber, err := s.ipamContext.dataStore.UnassignPodIPv4Address(ipamKey)
 	cidr := net.IPNet{IP: net.ParseIP(ip), Mask: net.IPv4Mask(255, 255, 255, 255)}
 	cidrStr := cidr.String()
-	if eni != nil && eni.AvailableIPv4Cidrs[cidrStr] != nil {
-		if s.ipamContext.enableIpv4PrefixDelegation && eni.AvailableIPv4Cidrs[cidrStr].IsPrefix == false {
+	if eni != nil {
+		if s.ipamContext.enableIpv4PrefixDelegation && eni.AvailableIPv4Cidrs[cidrStr] != nil && eni.AvailableIPv4Cidrs[cidrStr].IsPrefix == false {
 			log.Debugf("IP belongs to secondary pool with PD enabled so free IP from EC2")
 			s.ipamContext.tryUnassignIPFromENI(eni.ID)
-		} else if !s.ipamContext.enableIpv4PrefixDelegation && eni.AvailableIPv4Cidrs[cidrStr].IsPrefix == true {
+		} else if !s.ipamContext.enableIpv4PrefixDelegation && eni.AvailableIPv4Cidrs[cidrStr] == nil {
```

Comment:
Since this is a `DelNetwork` call, we know the IP/Prefix we just freed is attached to this particular ENI and so why should we check for internal data structure (i.e.,) `AvailableIPv4Cidrs[cidrStr]` to remove the IP/Prefix from the actual ENI? Shouldn't we always free/remove the IP/Prefix whenever there is a mismatch (sec ip in PD mode and prefix in sec ip mode)? Also, am assuming `eni.AvailableIPv4Cidrs[cidrStr] ` can never be nil for a just released IP? (mainly because we need it for cool down period check?)

Reply:
`cidrStr`  will be a /32 IP assigned to pod and that can be either from the secondary pool or prefix pool. With secondary IP mode `eni.AvailableIPv4Cidrs[cidrStr]` will not be `nil` but we need to free this IP only when we are in PD mode. Whereas `eni.AvailableIPv4Cidrs[cidrStr]` will be `nil` in PD mode since the key is `IP/28` and the prefix might be freeable in secondary mode. We can discuss this offline.

Reply:
Agree. I guess what I'm trying to say is that we should always free a secondary IP attached to an ENI in PD mode irrespective of whether `eni.AvailableIPv4Cidrs[cidrStr]` is nil or not and the same applies to a Prefix in secondary IP mode. Sure, let's take it offline.